### ISSUE TITLE
Add support for comparing meta tensors

### ISSUE THREAD

**Issue by pmeier** [pytorch/pytorch#68592]
## 🚀 Feature

See title.

## Motivation

Originally added as comment in #67032 :

https://github.com/pytorch/pytorch/blob/b1aa45a8a78211fcfec0923b5facca2b2829311d/torch/testing/_core.py#L63

@cbalioglu could you fill me in on what the behavior should be and the priority of this?

**Comment by mruberry** [comment]
What happens when `torch.testing._internal.TestCase.assertEqual()` is called on two meta tensors today?

**Comment by pmeier** [comment]
```python
>>> t = torch.empty((), device="meta")
>>> TestCase().assertEqual(t, t)
NotImplementedError: Could not run 'aten::isnan' with arguments from the 'Meta' backend. [...]
```

**Comment by cbalioglu** [comment]
Hey @pmeier, thanks for looking into this. As you already found out, it fails because some of the operators used in `assertEqual` do not support the meta backend. I raised my concern about this issue in the PR #67190 which I later closed because I thought that starting the conversation with an Issue might be an easier approach. Also see https://github.com/pytorch/pytorch/pull/67032#issuecomment-959852253 for the reason why I did not implement this feature in one of my earlier PRs.

In summary raising `NotImplementedError` in `assertEqual` simply skips (i.e. hides) bunch of device-specific unit tests that were in fact supposed to succeed or fail. They all surfaced when I "fixed" `assertEqual`, but since I had very little context about those tests, I left it to the Core team to come up with a more comprehensive fix.

**Comment by pmeier** [comment]
In that case let's wait for @ezyang's assessment of this and keep the current behavior for now.

**Comment by pmeier** [comment]
Actually, simply bailing out if we encounter a meta tensors is not quite equivalent to what we do today. `TestCase.assertEqual` does not handle meta tensors separately and thus fails with whatever error is raised when trying to access the data of a meta tensor. This "lazy detection" enables us to test some dispatch magic as is done here:

https://github.com/pytorch/pytorch/blob/0465f64bb88df6cca023c13884d2b33e649846ec/test/test_python_dispatch.py#L442

In this test we wrap a regular data tensor in meta tensor that dispatches all operators to the underlying data tensor. Thus, whenever we access data of the meta tensor, we actually access data of the regular tensor, which works fine with the current `TestCase.assertEqual`. This violates the assumption that we can never access data on a meta tensor. By bailing out early if we encounter a meta tensor we fail a test that would otherwise pass.

**Comment by ezyang** [comment]
Yes, data access is a big problem for meta tensors. Intuitively, you'd want something like "two meta tensors are equal if they are equal excluding their data", but if you do `(a == b).all()`

**Comment by pmeier** [comment]
@ezyang Special casing meta tensors to only compare the attributes is easy to achieve. See #72508. But note that this will fail at least one test as explained in https://github.com/pytorch/pytorch/issues/68592#issuecomment-993566803, since it builds on the assumption that `a == b` is always called even on meta tensors. cc @albanD

**Comment by ezyang** [comment]
Yes, adjusting tests will be necessary. Will be easiest if we can collect up what happens and look at it.
